### PR TITLE
Fix SAS7BDAT run-length encoding formula

### PR DESCRIPTION
**Pull request by jonashaag**
Run across an error/simplification? of a RLE formula in the SAS7BDAT parser. The original code https://bitbucket.org/jaredhobbs/sas7bdat/src/18cbd14407918c1aa90f136c1d6c5d83f307dba0/sas7bdat.py#lines-110:114 has the correct formula. I am not sure why we have another version.

Unfortunately I do not know how to add a regression test. The SAS file that reproduces the issue is proprietary, as is the SAS7BDAT file format, and I'm not aware of any means of creating a SAS file.

- [ ] closes #xxxx (Replace xxxx with the Github issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

**Comment by jonashaag** [comment]
Similar case here https://github.com/pandas-dev/pandas/issues/31243#issuecomment-686386759

**Comment by jonashaag** [comment]
pyreadstat has the same logic https://github.com/Roche/pyreadstat/blob/master/src/sas/readstat_sas_rle.c#L62-L63

File: pandas/io/sas/sas.pyx [hunk 1]
```diff
@@ -27,10 +27,8 @@ cdef const uint8_t[:] rle_decompress(int result_length, const uint8_t[:] inbuff)
         end_of_first_byte = <int>(inbuff[ipos] & 0x0F)
         ipos += 1
 
-        if control_byte == 0x00:
-            if end_of_first_byte != 0:
-                raise ValueError("Unexpected non-zero end_of_first_byte")
-            nbytes = <int>(inbuff[ipos]) + 64
+        if control_byte == 0x00 and ipos < length:
```
Comment (jonashaag) [inline pandas/io/sas/sas.pyx:30]:
```suggestion
        if control_byte == 0x00:
```

File: pandas/io/sas/sas.pyx [hunk 2]
```diff
@@ -28,9 +28,7 @@ cdef const uint8_t[:] rle_decompress(int result_length, const uint8_t[:] inbuff)
         ipos += 1
 
         if control_byte == 0x00:
-            if end_of_first_byte != 0:
-                raise ValueError("Unexpected non-zero end_of_first_byte")
-            nbytes = <int>(inbuff[ipos]) + 64
+            nbytes = <int>(inbuff[ipos] & 0xFF) + 64 + end_of_first_byte * 256
```
Comment (jonashaag) [inline pandas/io/sas/sas.pyx:31]:
```suggestion
            nbytes = <int>inbuff[ipos] + 64 + end_of_first_byte * 256
```

File: pandas/io/sas/sas.pyx [hunk 3]
```diff
@@ -28,9 +28,7 @@ cdef const uint8_t[:] rle_decompress(int result_length, const uint8_t[:] inbuff)
         ipos += 1
 
         if control_byte == 0x00:
-            if end_of_first_byte != 0:
-                raise ValueError("Unexpected non-zero end_of_first_byte")
-            nbytes = <int>(inbuff[ipos]) + 64
+            nbytes = <int>inbuff[ipos] + 64 + end_of_first_byte * 256
```
Comment (jonashaag) [inline pandas/io/sas/sas.pyx:31]:
```suggestion
            nbytes = <int>(inbuff[ipos]) + 64 + end_of_first_byte * 256
```

**Review by jreback — CHANGES_REQUESTED** [review]
can you add a test case / sample file that replicates the bug & a whatsnew note (I/O bug fixes for 1.5)

**Comment by jonashaag** [comment]
See my note above about a test case. Can add a what’s new

**Comment by jonashaag** [comment]
Unfortunately all test files I've seen so far (including the ones in pyreadstat) don't  reproduce the bug.

**Comment by jonashaag** [comment]
Unfortunately both of the files do not reproduce the bug

**Comment by jreback** [comment]
cc @bashtage WDYT?

**Comment by bashtage** [comment]
Would be good to have a file that requires this offset Is it possible to create one?

**Comment by jonashaag** [comment]
@bashtage Unfortunately I don't have a file that I can share and pyreadstat can't write SAS7BDAT files. Do you have any ideas where to get such a file?

**Comment by jonashaag** [comment]
@bashtage came up with a test file by zeroing almost all of the bytes of a real world file. PR now depends on https://github.com/pandas-dev/pandas/pull/47154

**Comment by bashtage** [comment]
> @bashtage came up with a test file by zeroing almost all of the bytes of a real world file. PR now depends on #47154

Once that is in rebase and then should be good.

**Comment by bashtage** [comment]
LGTM.

File: pandas/tests/io/sas/test_sas7bdat.py [hunk 4]
```diff
@@ -381,3 +381,10 @@ def test_exception_propagation_rle_decompress(tmp_path, datapath):
     tmp_file.write_bytes(data)
     with pytest.raises(ValueError, match="unknown control byte"):
         pd.read_sas(tmp_file)
+
+
+def test_0x00_control_byte(datapath):
+    # GH 47099
+    fname = datapath("io", "sas", "data", "0x00controlbyte.sas7bdat.bz2")
+    df = next(pd.read_sas(fname, chunksize=11_000))
+    assert df.shape == (11_000, 20)
```
Comment (bashtage) [inline pandas/tests/io/sas/test_sas7bdat.py:390]: Adding a review comment so I can approve.

**Review by bashtage — APPROVED** [review]
LGTM.

**Comment by jonashaag** [comment]
@jreback @mroeschke mind reviewing this?

File: doc/source/whatsnew/v1.5.0.rst [hunk 5]
```diff
@@ -940,6 +940,7 @@ I/O
 - Bug in :meth:`DataFrame.to_excel` when writing an empty dataframe with :class:`MultiIndex` (:issue:`19543`)
 - Bug in :func:`read_sas` with RLE-compressed SAS7BDAT files that contain 0x40 control bytes (:issue:`31243`)
 - Bug in :func:`read_sas` that scrambled column names (:issue:`31243`)
+- Bug in :func:`read_sas` with RLE-compressed SAS7BDAT files that contain 0x00 control bytes (:issue:`47099`)
```
[the remark this answers is not in the document]
Comment (jonashaag) [reply doc/source/whatsnew/v1.5.0.rst:943]: No it’s a different control byte

[the remark this answers is not in the document]
Comment (mroeschke) [reply doc/source/whatsnew/v1.5.0.rst:943]: Ah thanks missed that 😅

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @jonashaag

**Comment by jonashaag** [comment]
Thanks :)

**Pull request merged**